[CLEANUP] Extraction sélection des épreuves de certification (PIX-992)

P-Jeremy opened the pull request:
## :unicorn: Problème
La sélection des épreuves de certification se faisait par le placement profile service. ça n'est pas la responsabilité du placement profile.

Suite du #1675 

## :robot: Solution
Déplacer algorithme de sélection dans le certification challenge service.

## :rainbow: Remarques
A SUIVRE: 
- Revoir le retour de `retrieve-or-create-certification-course` pour ne pas retourner tout l'assessment ainsi que toutes les questions alors que le serializer n'a besoin que de l'`assessmentId `et du nombre de questions.
- Est-ce que les champs `examinerComment` et `asSeenEndScreen` sont pertinent dans l'objet `CertificationCourse` (on utilise ce même objet pour la création **et** le jury => est-ce judicieux ?)


## :100: Pour tester

Générer plein de tests de certifs avant et après refacto et comparer les résultats.

```
$ PGSSLMODE=require USER_COUNT=1000 node scripts/generate-certification-test-statistics.js
```

sroccaserra committed:
[skip ci] Start the analysis

Giroteti committed:
STEP 1 : implement CertificationCourse.from()

Giroteti committed:
Add UserCompetence domain builder

Giroteti committed:
STEP 2 : implement CertificationCourse.from()

Giroteti committed:
STEP 3 : implement CertificationCourse.from()

P-Jeremy committed:
Add _saveCertificationCourse

sroccaserra committed:
STEP 4: extract userCompetencesWithChallenges

Giroteti committed:
generateCertificationChallenges : challenges input

Giroteti committed:
PickCertificationChallenges returns challenges

aceol committed:
Move pickCertificationChallenges to certificationChallengesService

sroccaserra committed:
Made generateCertificationChallenges private

P-Jeremy committed:
rm generateCertificationChallenges in pickCertificationChallenges

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr1694.review.pix.fr
- Orga: https://orga-pr1694.review.pix.fr
- Certif: https://certif-pr1694.review.pix.fr
- Admin: https://admin-pr1694.review.pix.fr
- API: https://api-pr1694.review.pix.fr/api/

 Please check it out!

sroccaserra committed:
Adapt script to generate certification profiles

Giroteti committed:
Do not mutate UserCompetences.challenges

Giroteti committed:
Explained the 2nd check on CertifCourse existence

The certification course creation goes like this :

Check certification course existence
if (!exist) {
 Build certification course
 Recheck certification course existence (**Why ? we already checked it above**)
 if (exist) {
   return existing one
 } else {
   save new one and return it
 }
}

The "Build certification course" part is costly and slow so that it is likely that two calls might enter it simultaneously, at the end of this statement, we want the slowest of them to return the result of the fastest of them.

sroccaserra committed:
Refactor certificationChallengeRepository test

Giroteti committed:
CertificationCourseRepo saves Challenges as well

Giroteti committed:
Assessment.createForAssessment factory method

sroccaserra commented on api/lib/domain/models/Assessment.js:
```diff
@@ -108,6 +108,15 @@ class Assessment {
     return this.isCompetenceEvaluation() || this.isForCampaign();
   }
+
+  static createForCertificationCourse({ userId, courseId }) {
+    return new Assessment({
+      userId,
+      courseId,
+      state: Assessment.states.STARTED,
+      type: Assessment.types.CERTIFICATION,
+    });
+  }
```
```suggestion
  static createForCertificationCourse({ userId, certificationCourseId }) {
    return new Assessment({
      userId,
      certificationCourseId,
      state: Assessment.states.STARTED,
      type: Assessment.types.CERTIFICATION,
    });
  }
```

laura-bergoens replied:
1) J'ai vu que c'est écrit dans le descriptif de PR, qu'il est prévu d'imaginer un nouveau modèle Assessment qui va contenir le strict nécessaire, donc c'est cool !
2) Le suggested change de @sroccaserra est TRES IMPORTANT :
- Le champ courseId est rempli lorsque l'assessment est lié à une Démo (courseId est une chaîne contenant l'id de la démo). Exemple, si vous allez sur le site de [Pix](https://pix.fr/), vous pouvez "Essayer sans inscription". En cliquant, vous êtes redirigé vers https://app.pix.fr/api/courses/rec8rnpOgmfZyd2Ng  (rec8rnpOgmfZyd2Ng est l'id de la démo). Ca va conséquemment créer un assessment dont le courseId va être rec8rnpOgmfZyd2Ng !
- Le champ `certificationCourseId` est bien une clé étrangère vers un `certification-course`

Giroteti replied:
@laura-bergoens Est-ce que du coup "courseId" aurait plus de sens en "demoCourseId" ou alors il sert aussi à autre chose et c'est trop restrictif ?

sroccaserra commented on api/lib/domain/usecases/retrieve-last-or-create-certification-course.js:
```diff
@@ -36,80 +39,65 @@ module.exports = async function retrieveLastOrCreateCertificationCourse({
+    domainTransaction
+  });
+
+  const newAssessment = Assessment.createForCertificationCourse({ userId, courseId: savedCertificationCourse.id });
```
```suggestion
  const newAssessment = Assessment.createForCertificationCourse({ userId, certificationCourseId: savedCertificationCourse.id });
```

sroccaserra commented on api/tests/unit/domain/usecases/retrieve-last-or-create-certification-course_test.js:
```diff
@@ -216,27 +231,26 @@ describe('Unit | UseCase | retrieve-last-or-create-certification-course', () =>
           const mockAssessment = {
             userId,
-            certificationCourseId: savedCertificationCourse.id,
+            courseId: savedCertificationCourse.id,
```
```suggestion
            certificationCourseId: savedCertificationCourse.id,
```

Anne-Gaelle-S commented on api/lib/domain/usecases/retrieve-last-or-create-certification-course.js:
```diff
@@ -60,51 +64,48 @@ async function _startNewCertification({
     throw new UserNotAuthorizedToCertifyError();
   }
 
+  // certficationProfile = CertificationProfile.from(placementProfile);
```
Reste un commentaire (peut être enlevé par la suite ?)

Anne-Gaelle-S commented on api/lib/domain/usecases/retrieve-last-or-create-certification-course.js:
```diff
@@ -60,51 +64,48 @@ async function _startNewCertification({
+
   // FIXME : Ce n'est pas la responsabilité du placementProfile que d'avoir les challenges
   placementProfile = await placementProfileService.fillPlacementProfileWithChallenges(placementProfile);
+  // certificationCourseChallenges = identifierLesChallenges(certificationProfile); // $$$
```
ici aussi commentaire

Anne-Gaelle-S commented on api/lib/domain/usecases/retrieve-last-or-create-certification-course.js:
```diff
@@ -60,51 +64,48 @@ async function _startNewCertification({
+
+  const newCertificationCourse = CertificationCourse.from({ certificationCandidate });
+  // const certificationCourse = CertificationCourse.from(certificationProfile, certificationCourseChallenges, certificationCandidate);
+  // certificationCourseRepository.save(certificationCourse);
```
com

Anne-Gaelle-S commented on api/lib/domain/services/certification-challenges-service.js:
```diff
@@ -4,16 +4,18 @@ const CertificationChallenge = require('../models/CertificationChallenge');
-        });
+    const challenges = _.flatMap(userCompetences, (userCompetence) => {
+      return userCompetence.challenges;
+    });
```
Je crois qu'on peut l'écrire aussi comme ça 
```suggestion
    const challenges = _(userCompetences).flatMap(uc => uc.challenges)
```
ou encore
```suggestion
   const challenges = _.flatMap(userCompetences, (uc) => uc.challenges);
```

Anne-Gaelle-S commented on api/tests/unit/domain/usecases/retrieve-last-or-create-certification-course_test.js:
```diff
@@ -218,9 +218,20 @@ describe('Unit | UseCase | retrieve-last-or-create-certification-course', () =>
+          const challenge1 = domainBuilder.buildCertificationChallenge();
+          const challenge2 = domainBuilder.buildCertificationChallenge();
+          const generatedCertificationChallenges = [challenge1, challenge2];
+          const savedCertificationChallenge1 = {
+            id: 'savedCertificationChallenge1',
+          };
+          const savedCertificationChallenge2 = {
+            id: 'savedCertificationChallenge2',
+          };
```
```suggestion
          const savedCertificationChallenge1 = { id: 'savedCertificationChallenge1' };
          const savedCertificationChallenge2 = { id: 'savedCertificationChallenge2' };
```

Anne-Gaelle-S commented on api/tests/unit/domain/usecases/retrieve-last-or-create-certification-course_test.js:
```diff
@@ -151,10 +152,21 @@ describe('Unit | UseCase | retrieve-last-or-create-certification-course', () =>
+
         beforeEach(() => {
-          placementProfile = { isCertifiable: sinon.stub().returns(true), userCompetences: 'someUserCompetences' };
+          // TODO : use the domainBuilder to instanciate userCompetences
```
Reste un TODO ici, peut être enlevé par la suite ?

Anne-Gaelle-S reviewed (commented):
C'est assez dur de faire la revue commit par commit, je vais tenter de faire la revue générale plutôt

Anne-Gaelle-S commented on api/lib/infrastructure/repositories/certification-course-repository.js:
```diff
@@ -8,14 +8,24 @@ const DomainTransaction = require('../DomainTransaction');
 
   async save({ certificationCourse, domainTransaction = DomainTransaction.emptyTransaction() }) {
-    const certificationCourseToSave = _adaptModelToDb(certificationCourse);
+    const certificationCourseToSaveDTO = _adaptModelToDb(certificationCourse);
```
C'est quoi DTO ?

Giroteti replied:
Data Transfer Object => c'est un objet donc la raison d'exister est de faire du mapping entre une source et une destination, ici entre un objet du domain `CertificationCourse` et la table certification-courses. C'est en général un objet sans méthode, directement sérialisable, qui contient que des primitives du language, relativement plat. 
T'as souvent ça au moment des entrée/sorties (enregistrement en DB, récupération depuis la DB, avant un appel réseau pour générer la requête, après un appel réseau pour interpréter la réponse, etc).

Anne-Gaelle-S commented on api/tests/integration/infrastructure/repositories/certification-course-repository_test.js:
```diff
@@ -26,22 +26,52 @@ describe('Integration | Repository | Certification Course', function() {
     it('should persist the certif course in db', async () => {
       // when
-      await certificationCourseRepository.save({ certificationCourse });
+      const savedCertificationCourse = (await certificationCourseRepository.save({ certificationCourse }));
```
```suggestion
      const savedCertificationCourse = await certificationCourseRepository.save({ certificationCourse });
```

Anne-Gaelle-S commented on api/tests/integration/infrastructure/repositories/certification-course-repository_test.js:
```diff
@@ -26,22 +26,52 @@ describe('Integration | Repository | Certification Course', function() {
+        'completedAt',
+        'createdAt',
+      ];
+      const fieldsToOmitInCertificationChallenge = [
+        'id',
+        'courseId'
+      ];
```
```suggestion
      const fieldsToOmitInCertificationChallenge = [ 'id', 'courseId' ];
```

Anne-Gaelle-S reviewed (approved):
Des petites suggestions de style. 
Bravo pour le travail ! 🎉 
Je n'ai pas encore testé fonctionnellement par contre

laura-bergoens commented on api/lib/domain/models/CertificationCourse.js:
```diff
@@ -41,6 +41,20 @@ class CertificationCourse {
   }
+
+  static from({ certificationCandidate, challenges }) {
+    return new CertificationCourse({
```
Du coup on fait le même constat que pour `Assessment`, ça pourrait être intéressant de lui trouver son petit modèle dédié + validation. Pas sur cette PR évidemment, mais on prend note **(PAS BLOQUANT)**

laura-bergoens commented on api/lib/domain/services/certification-challenges-service.js:
```diff
@@ -1,19 +1,93 @@
-          courseId: certificationCourseId,
-        });
+  async pickCertificationChallenges(placementProfile) {
+    const knowledgeElementsByCompetence = await knowledgeElementRepository
+      .findUniqByUserIdGroupedByCompetenceId({ userId: placementProfile.userId, limitDate: placementProfile.profileDate });
```
Le besoin de grouper les knowledge-elements par compétence est legacy je pense.
Nul besoin de le faire ici (en plus, ça ferait gagner un peu de temps, on crache pas dessus).
Je propose d'appeler `knowledgeElementRepository.findUniqByUserId` suivi d'un `KnowledgeElement.findDirectlyValidated`  .   `KnowledgeElement.findDirectlyValidatedFromGroups` est une méthode créée uniquement pour ce contexte, donc on peut la changer sans problème (on verra d'ailleurs que cette méthode fait actuellement un flat() d'une collection péniblement groupée quelques lignes auparavant !) **(BLOQUANT)**

laura-bergoens commented on api/lib/domain/services/certification-challenges-service.js:
```diff
@@ -1,19 +1,93 @@
+    const challengesAlreadyAnswered = challengeIdsCorrectlyAnswered.map((challengeId) => Challenge.findById(allChallenges, challengeId));
+
+    challengesAlreadyAnswered.forEach((challenge) => {
+      if (!challenge) {
+        return;
+      }
```
A vérifier, mais ce `if` est, je pense, très très superflu. Je doute que `challengeRepository.findOperative()` puisse renvoyer des challenges qui n'existent pas :thinking: **(BLOQUANT)**

laura-bergoens commented on api/lib/domain/services/certification-challenges-service.js:
```diff
@@ -1,19 +1,93 @@
+    const challengeIdsCorrectlyAnswered = await answerRepository.findChallengeIdsFromAnswerIds(answerIds);
+
+    const allChallenges = await challengeRepository.findOperative();
+    const challengesAlreadyAnswered = challengeIdsCorrectlyAnswered.map((challengeId) => Challenge.findById(allChallenges, challengeId));
```
Je vous laisse juger de décider si cet appel à la statique a un grand intérêt de lisibilité plutôt que :
```suggestion
    const challengesAlreadyAnswered = challengeIdsCorrectlyAnswered.map((challengeId) => allChallenges.find((challenge) => challenge.id === challengeId));
```
**(PAS BLOQUANT)**

laura-bergoens commented on api/lib/domain/services/certification-challenges-service.js:
```diff
@@ -1,19 +1,93 @@
+        return;
+      }
+
+      const userCompetence = _getUserCompetenceByChallengeCompetenceId(placementProfile.userCompetences, challenge);
+
+      if (!userCompetence || !userCompetence.isCertifiable()) {
+        return;
+      }
```
A la place de tout ce morceau je propose :
```js
if (!placementProfile.isCompetenceCertifiable(challenge.competenceId)) {
  return ;
}
```
Et dans la classe PlacementProfile une méthode pour faire le boulot (qu'on pourra tester unitairement) :
```js
isCompetenceCertifiable(competenceId) {
  const userCompetence = this.userCompetences.find((uc) => uc.id === competenceId);
  return userCompetence && userCompetence.isCertifiable();
}
```
**(BLOQUANT)**

laura-bergoens commented on api/lib/domain/services/certification-challenges-service.js:
```diff
@@ -1,19 +1,93 @@
+
+      challenge.skills
+        .filter((skill) => _skillHasAtLeastOneChallenge(skill, allChallenges))
+        .forEach((publishedSkill) => userCompetence.addSkill(publishedSkill));
```
En vrai il faudrait aussi déresponsabiliser `userCompetence` (via le retrait de méthodes telles que `addSkill` ou `orderSkillsOfCompetenceByDifficulty`, mais la PR commence à être grosse alors ça peut être fait dans un deuxième temps
**(PAS BLOQUANT)**

laura-bergoens replied:
On pourrait aussi imaginer pouvoir se priver de la méthode statique `orderSkillsOfCompetenceByDifficulty` en faisant en sorte que les acquis soit naturellement triés par ordre de difficulté lors de l'ajout `addSkill`
**(BLOQUANT)**

laura-bergoens commented on api/lib/domain/services/certification-challenges-service.js:
```diff
@@ -1,19 +1,93 @@
+          const challengesLeftToAnswer = _.difference(challengesToValidateCurrentSkill, challengesAlreadyAnswered);
+
+          const challengesPoolToPickChallengeFrom = (_.isEmpty(challengesLeftToAnswer)) ? challengesToValidateCurrentSkill : challengesLeftToAnswer;
+          const challenge = _.sample(challengesPoolToPickChallengeFrom);
```
on en profiterait pas pour faire un "random" maîtrisé / avec seed lié à l'utilisateur ? **(PAS BLOQUANT)**

Giroteti replied:
le seed c'est pas possible en JS sans utiliser une lib qu'on a pas encore, on va étudier le truc...

laura-bergoens commented on api/lib/domain/services/certification-challenges-service.js:
```diff
@@ -1,19 +1,93 @@
+          const challengesPoolToPickChallengeFrom = (_.isEmpty(challengesLeftToAnswer)) ? challengesToValidateCurrentSkill : challengesLeftToAnswer;
+          const challenge = _.sample(challengesPoolToPickChallengeFrom);
+
+          const certificationChallenge = new CertificationChallenge({
+            challengeId: challenge.id,
+            competenceId: skill.competenceId,
+            associatedSkillName: skill.name,
+            associatedSkillId: skill.id
+          });
```
:pray: **(SUPER COOL)**

laura-bergoens commented on api/lib/infrastructure/repositories/certification-course-repository.js:
```diff
@@ -8,14 +8,24 @@ const DomainTransaction = require('../DomainTransaction');
-    return _toDomain(savedCertificationCourse);
+    const savedCertificationCourseDTO = await new CertificationCourseBookshelf(certificationCourseToSaveDTO).save(null, options);
+
+    // FIXME : Save all challenges in a single DB request
```
Vous le faites dans une PR différente?**(PAS BLOQUANT)**

laura-bergoens commented on api/lib/infrastructure/repositories/certification-course-repository.js:
```diff
@@ -8,14 +8,24 @@ const DomainTransaction = require('../DomainTransaction');
+    const savedCertificationCourseDTO = await new CertificationCourseBookshelf(certificationCourseToSaveDTO).save(null, options);
+
+    // FIXME : Save all challenges in a single DB request
+    const savedChallenges = await Promise.all(certificationCourse.challenges.map((certificationChallenge) => {
+      const certificationChallengeWithCourseId = { ...certificationChallenge, courseId: savedCertificationCourseDTO.id };
+      return certificationChallengeRepository.save({ certificationChallenge: certificationChallengeWithCourseId, domainTransaction });
+    }));
```
Ne pas faire de `Promise.all`, ce n'est pas sain pour une situation prod-like, here's why :
avec `Promise.all`, on va lancer plein de `save` en même temps (éventuellement jusqu'à : nbCompetences * nbQuestions, donc 16 * 3 = 48 en même temps!). Ce qui veut dire que potentiellement votre seul usecase va réserver, à lui tout seul, plusieurs connexions BDD. C'est pas cool pour les autres requêtes effectuées au sein du même container.

BIG EDIT : (je laisse le commentaire au dessus car on apprend quand même des choses), mais en fait il n'y aura bien qu'une seule connexion BDD sollicitée (car on passe par une transaction). Après bon, on peut quand même chipoté et dire que les faire de façon séquentielle c'est aussi acceptable.
Cela dit, si c'est prévu de toutes les enregistrer d'un coup, alors j'arrête de vous embêter :cat: 

**(PAS BLOQUANT)**

laura-bergoens commented on api/lib/infrastructure/repositories/certification-course-repository.js:
```diff
@@ -8,14 +8,24 @@ const DomainTransaction = require('../DomainTransaction');
+
+    // FIXME : Save all challenges in a single DB request
+    const savedChallenges = await Promise.all(certificationCourse.challenges.map((certificationChallenge) => {
+      const certificationChallengeWithCourseId = { ...certificationChallenge, courseId: savedCertificationCourseDTO.id };
```
Si vous êtes motivés, renommez `courseId` en `certificationCourseId` dans le modèle `CertificationChallenge`, c'est une abérration et courseId fait référence à autre chose dans d'autres contextes donc c'est confus pour rien !
**(PAS BLOQUANT)**

laura-bergoens reviewed (changes requested):
Je requête des changements, mais franchement je suis ravie que ce soit fait. Ce usecase avait besoin d'un petit coup de balai. Pingez moi si vous avez des questions OU si vous voulez qu'on voit les commentaires ensemble OU si vous avez tout pris en compte et qu'il faut que je repasse ma revue.

Je vous aime :heart: , groove is in the heart

EDIT: j'ai qualifié les commentaires en **(PAS BLOQUANT)** ou **(BLOQUANT)**
Bloquant ça veut surtout dire que j'aimerais bien avoir une réponse, pas forcément que la suggestion soit appliquée

laura-bergoens reviewed (approved):
Vu avec @EGI-OCTO 
Le déplacement tel qu'il est fait là me paraît ISO par rapport à la version d'avant.
L'ensemble des commentaires que j'ai mis sur cette PR sont à garder en TODO List pour les PRs successives de refacto.
